### ISSUE TITLE
Confusing failures when using PowerShell < 5.0

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#27938]
If you run on Windows with PoSh < 5 you hit this code which prints a message and quits:

https://github.com/flutter/flutter/blob/e99c881f8d0e63042af6e2a3f6182c3ca6967442/bin/internal/update_dart_sdk.ps1#L30-L32

However, since this is inside update_dart_sdk.it only quits that file, it doesn't stop what it's doing. The parent script then just continues on and tries to run `pub`:

https://github.com/flutter/flutter/blob/e99c881f8d0e63042af6e2a3f6182c3ca6967442/bin/flutter.bat#L84-L85

Which then outputs:

```
Building flutter tool...
Running pub upgrade...
The system cannot find the path specified.
Error Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (9 tries
 left)

Waiting for 2 seconds, press CTRL+C to quit ...
```

This is a little confusing, since it looks like a failure to run `pub`, when in reality the Dart SDK was never downloaded because we bailed because of the earlier error. We should probably make update_dart_sdk quit with an error code, and then handle that in the parent flutter.bat.

**Comment by kangwang1988** [comment]
Yes. It would be better.

**Comment by Cedar7** [comment]
I have faced the same error today, after updating my PowerShell found at the following page
[(https://docs.microsoft.com/en-us/powershell/scripting/install/installing-windows-powershell?view=powershell-6)], I have managed to fix the error.

**Comment by ArtemGr** [comment]
The `PowerShell $psMajorVersionRequired.0 or newer` is really confusing because the latest PowerShell is 7 at https://github.com/PowerShell/PowerShell/releases but Flutter won't work with it, at least on Windows 7 it won't.

Like @Cedar7 said, one needs to install the PowerShell 5 specifically and that from a Windows Update package (https://docs.microsoft.com/en-us/powershell/wmf/setup/install-configure).

**Comment by DanTup** [comment]
> the latest PowerShell is 7 at https://github.com/PowerShell/PowerShell/releases but Flutter won't work with it, at least on Windows 7 it won't

My guess is that this is PowerShell Core? That's not supported, see https://github.com/flutter/flutter/issues/18790#issuecomment-400196934. I think we should also detect this and show a sensible error though.

Let me know if seeing errors on PowerShell >= 5.0 that is *not* PowerShell Core though.

**Comment by DanTup** [comment]
Well it seems PowerShell versioning is much more confusing than I realised. There were two types of PowerShell... "PowerShell" (shipped with Windows, supports everything) and "PowerShell Core" (a cross-platform version that supports less). PowerShell ships with windows (and its edition is "Desktop") and the other you can download from GitHub.

However, despite being called "PowerShell Core" on GitHub, the installer for it just calls it "PowerShell" and the entries in the start menu just say "PowerShell 6".

That said, `Unblock-File` was added to PowerShell core (it works in the current stable v6 release) so maybe it's reasonable to say that is supported now.

@ArtemGr if you open a new issue with details of what you see fail when using PowerShell (Core) I'll see if I can repro (I don't have easy access to Windows 7, but the error message might give me clues if whether that's relevant).

**Comment by ArtemGr** [comment]
> ArtemGr if you open a new issue with details of what you see fail when using PowerShell (Core)

The problem with PowerShell (Core) is that `flutter` fails with `Flutter requires PowerShell 5.0 or newer`.  
I figured that I need the other PowerShell in the end, but it took me a couple days to do so.  
My suggestion hence is to clarify that error message, pointing the users at the right PowerShell.

**Comment by DanTup** [comment]
@ArtemGr I used PowerShell core v6 today but didn't have that issue. Could you post the output of running `$PSVersionTable` in the window where you're seeing this failure and I'll take a look.

**Comment by ArtemGr** [comment]
Sure, I'll try to uninstall the KB3191566 and then I'll maybe see the failure again.

**Comment by DanTup** [comment]
@ArtemGr that updates PowerShell (non-core), so I think the failure you saw was probably from running in PowerShell desktop <= 5.0. If you install PowerShell Core you end up with two, one called "PowerShell" and one called "PowerShell 6" (or 7). My guess is that you had run Flutter from the Desktop (non-core) edition.

**Comment by ArtemGr** [comment]
> My guess is that you had run Flutter from the Desktop (non-core) edition.

Of course I did. What's the point of your previous question - https://github.com/flutter/flutter/issues/27938#issuecomment-523550749 - otherwise?

**Comment by ArtemGr** [comment]
@DanTup, So I was wrong about the error message on PowerShell 7, sorry about that.
Here's the actual one:
```
PS C:\spool> flutter
Checking Dart SDK version...
Windows PowerShell завершена из-за следующей ошибки:
 Не удалось загрузить тип "System.Management.Automation.Runspaces.EarlyStartup"
из сборки "System.Management.Automation, Version=3.0.0.0, Culture=neutral, Publi
cKeyToken=31bf3856ad364e35".
Error: Unable to update Dart SDK. Retrying...
```

![powershell-7](https://user-images.githubusercontent.com/41847/63455371-56db7d00-c455-11e9-9688-14c912a10537.png)

**Comment by ArtemGr** [comment]
```
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.0-preview.2
PSEdition                      Core
GitCommitId                    7.0.0-preview.2
OS                             Microsoft Windows 6.1.7601 Service Pack 1
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0:}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by DanTup** [comment]
That error looks like it's coming from PowerShell itself, though since that's still a preview version so it may be expected that some things don't work. Do you see the same issue if you use the [latest stable version (6.2.2)](https://github.com/PowerShell/PowerShell/releases/tag/v6.2.2)?

It may be worth opening an issue at https://github.com/PowerShell/PowerShell/issues about it (feel free to CC me in case I can be of any help in tracking it down).

**Comment by ArtemGr** [comment]
> Do you see the same issue if you use the latest stable version (6.2.2)?

PowerShell Core 6.2.2 had a problem installing.

> It may be worth opening an issue at https://github.com/PowerShell/PowerShell/issues about it

I don't need PowerShell Core, I've only reinstalled it because you've asked. All I need here is to run the `flutter` command, and my concern is that the error message is currently ambiguous, leading the user towards trying the untested PowerShell Core setups (and failing at that). Would be nice if the error message and/or the installation article would clarify that PowerShell Core is not a seamless upgrade from PowerShell 4 or 5.

**Comment by DanTup** [comment]
I'm not sure I understand. Flutter works with PowerShell 5 (which is Desktop edition) and PowerShell 6 (which is Core edition) in my testing. The message tells the user they need at least PowerShell 5, which I believe is accurate. If PowerShell 7 (Core) gets a stable release and Flutter does not work in it, then I think it's fair to consider that a bug.

The version you tried (PowerShell 7 (Core edition)) is a preview version and is throwing errors inside its commands - these seems highly likely a PowerShell issue, and I'm not sure it's necessary for the Flutter message to call out that preview versions of dependent software may have issues.

**Comment by ArtemGr** [comment]
I see your point. But fact is, I've spent a lot of time and for a while considered Flutter broken because PowerShell 6 and 7 won't work for me on Windows 7 (I wonder if this is the OS you're testing on; on Windows 10 PowerShell installation is a non-issue, it is already there). And I've come here in hopes of making this experience less hurtfull for the rest of the Flutter users who might find themselves in a similar situation. You've mentioned several times already here on github that PowerShell 6 is not supported. Now you're saying that you've tested it and it works fine and hence everything's cool. Which is it, is PowerShell 6 supported or not? If it is supported, do you have CI runs that verify Flutter installation with it on Windows 7? As of now we only have information from two people, you and me, so there's 50/50 probability that PowerShell 6 will work for other users. Is it good enough from the user experience perspective?

**Comment by DanTup** [comment]
> You've mentioned several times already here on github that PowerShell 6 is not supported. Now you're saying that you've tested it and it works fine and hence everything's cool. Which is it, is PowerShell 6 supported or not?

When I wrote the first comments about this, PowerShell 6 (which was generally referred to as "PowerShell Core") was missing functionality that PowerShell 5 (generally referred to as "PowerShell") had. The PowerShell team said they had no intention to implement those features, which meant Flutter would fail. At the time, "PowerShell Core" was not supported, and "PowerShell" and "PowerShell Core" seemed like two distinct products.

They apparently changed their minds on that, and the current stable version of PowerShell Core now supports those features, and therefore it works fine for Flutter. I guess the intention was to make the change from PowerShell 5 (Desktop) -> PowerShell 6 (Core) easier - though I'm not kept all that up-to-date on PowerShell lately.

> As of now we only have information from two people, you and me, so there's 50/50 probability that PowerShell 6 will work for other users.

From what you're said, you were unable to install PowerShell 6. This is not the same as "Flutter doesn't work on PowerShell 6". If you can get PowerShell 6 to install, and Flutter fails - I can help try and debug it. However I'm not well placed to help debug PowerShell 6 not installing, nor internal errors in PowerShell 7 - that's why I asked you to open an issue in the PowerShell repo - and I asked you to CC me there, so I can assist if possible. I'm entirely open to helping - but PowerShell bugs and PowerShell installation issues aren't fixable by Flutter - and so far there hasn't been any information showing that Flutter fails to work on the latest stable version of PowerShell Core. If any is provided, I am more than happy to try and resolve the problem.

> I'm not sure whether you realize that PowerShell 7 is the version most users will get when looking for a fresh PowerShell release with Google.

It's unfortunate that GitHub doesn't do a good job of letting you filter these, and that PowerShell have huge changelogs that push the stable release down the releases page, but that's unfortunately not something Flutter can fix. I'm not sure what you're proposing - if the PowerShell preview version is broken, it needs to be reported to the PowerShell team. Again - I'm happy to be CC'd and try to help find a resolution.

**Comment by ArtemGr** [comment]
> I'm not sure what you're proposing

I'm proposing the same thing over and over ( from my first message - https://github.com/flutter/flutter/issues/27938#issuecomment-523154750 - and through several messages after ): to clarify in the Flutter error message and on the flutter Windows installation page that Flutter has been tested to work with PowerShell 5 and 6. The current error message, `Flutter requires PowerShell $psMajorVersionRequired.0 or newer` incorrect.

**Comment by ArtemGr** [comment]
Take semver for example. If you specify a version >5 then it will include versions 7, 8, 9, etc, even if they aren't stable. If you want a stable version, you'd want to add an upper boundary.

**Comment by DanTup** [comment]
The problem with updating it to list the specific stable versions of PowerShell is that the message will quickly get out of date. Flutter only ships to the stable channel periodically, so suggesting PowerShell 7 is not supported will be confusing when a) the issue you're hitting is fixed or b) PowerShell 7 goes stable and it is intended to be supported. It could take many months for the updated message to reach the stable channel.

I don't think saying "at least version x" of any software implies that there will be no issues if you install some unstable preview release of x. It's implied when you install a preview version of something that it may have issues with it. If you do have issues with it, the best thing is to raise them against the software so that they can be fixed, rather than all dependent software doing specific version checks against specific preview versions (for all we know, the next preview version may fix it).

Is there a reason you don't wish to open an issue in the PowerShell repo? I've offered to help where I can if you CC me on it. It would be much better to get the issue fixed than try to maintain a mapping of which PowerShell preview versions are known not to work in Flutter.

**Comment by ArtemGr** [comment]
> I don't think saying "at least version x" of any software implies that there will be no issues if you install some unstable preview release of x.

There's a simple rule coming from Kent Beck's era: "if it isn't tested, it doesn't work".
There **will** be issues with untested versions, stable or not, that's why you ought to specify the tested versions of the dependencies on the stable channel.
Unbound versions on a stable channel are not okay.
